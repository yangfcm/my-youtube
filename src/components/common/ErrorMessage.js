import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const renderError = (error) => {
  return (
    <div className="text-danger text-center my-3 py-3">
      <FontAwesomeIcon icon="exclamation-triangle" className="mx-2" />
      {error.displayMessage}
    </div>
  );
};

const ErrorMessage = ({ error }) => {
  // console.log(process.env);
  if (process.env.NODE_ENV === "development") {
    console.log(error);
  }
  return renderError(error);
};

export default ErrorMessage;
