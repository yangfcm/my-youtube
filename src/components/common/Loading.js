import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Loading = () => {
  return (
    <div className="text-center my-3 py-3">
      <FontAwesomeIcon icon="spinner" className="mx-2" />
      Loading...
    </div>
  );
};

export default Loading;
