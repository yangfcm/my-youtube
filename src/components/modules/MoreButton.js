import React from "react";

const MoreButton = ({ onClickMore, children }) => {
  return (
    <div className="text-center">
      <button
        className="btn btn-light"
        style={{ width: "100%" }}
        onClick={onClickMore}
      >
        <span className="text-dark">{children ? children : "More..."}</span>
      </button>
    </div>
  );
};

export default MoreButton;
