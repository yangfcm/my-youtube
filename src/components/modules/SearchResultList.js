import React from "react";
import SearchItem from "./SearchItem";

const SearchResultList = ({ searchResultList }) => {
  return (
    <div>
      {searchResultList.map((result, index) => {
        return <SearchItem key={index} item={result} />;
      })}
    </div>
  );
};

export default SearchResultList;
