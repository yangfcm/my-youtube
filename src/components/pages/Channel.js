import React from "react";
import Banner from "../layout/Banner";
import Menu from "../layout/Menu";
import { mainMenuItems } from "../../settings";

class Channel extends React.Component {
  render() {
    return (
      <div>
        <Banner />
        <Menu menuItems={mainMenuItems} />
        Channels page
      </div>
    );
  }
}

export default Channel;
