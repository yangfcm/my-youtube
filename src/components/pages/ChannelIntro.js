import React from "react";
import moment from "moment";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Loading from "../common/Loading";
import ErrorMessage from "../common/ErrorMessage";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { fetchChannelIntro, clearError } from "../../actions/app";
import { separateNumber } from "../../utils/helper";

class ChannelIntro extends React.Component {
  state = {
    channelIntro: null,
    error: null,
  };
  componentDidMount = async () => {
    const { channelId } = this.props;
    await this.props.fetchChannelIntro(channelId);
    if (this.props.error) {
      this.setState({
        error: this.props.error,
      });
      return;
    }
    if (this.props.channel) {
      if (this.props.channel.items && this.props.channel.items[0]) {
        this.setState({
          channelIntro: this.props.channel.items[0],
        });
      } else {
        this.setState({
          error: "The channel doesn't exist.",
        });
      }
    }
  };
  componentWillUnmount = () => {
    this.props.clearError();
  };
  render() {
    // console.log(this.state);
    const { channelIntro } = this.state;
    return (
      <React.Fragment>
        {!this.state.error && !this.state.channelIntro && <Loading />}
        {this.state.error && <ErrorMessage message={this.state.error} />}
        {!this.state.error && this.state.channelIntro && (
          <div className="mt-3 row justify-content-center">
            <div className="col-md-10 col-lg-9">
              <div className="d-flex flex-column flex-sm-row">
                <div className="mr-3 my-3 text-center">
                  <img
                    className="rounded-circle"
                    style={{ maxWidth: "120px", height: "auto" }}
                    src={channelIntro.snippet.thumbnails.medium.url}
                    alt={channelIntro.snippet.title}
                  />
                </div>
                <div className="flex-grow-1">
                  <div className="d-flex justify-content-center flex-column">
                    <h3 className="my-3 text-center font-weight-bold text-primary">
                      {channelIntro.snippet.title}
                    </h3>
                    <ul className="list-group">
                      <li className="list-group-item">
                        <FontAwesomeIcon icon="calendar-day" /> Published on{" "}
                        {moment(channelIntro.snippet.publishedAt).format(
                          "D MMM YYYY"
                        )}
                      </li>
                      {channelIntro.snippet.description && (
                        <li className="list-group-item">
                          {" "}
                          <FontAwesomeIcon icon="envelope-open-text" />{" "}
                          Description - {channelIntro.snippet.description}
                        </li>
                      )}
                      <li className="list-group-item">
                        <FontAwesomeIcon icon="user-friends" />{" "}
                        {separateNumber(
                          channelIntro.statistics.subscriberCount
                        )}{" "}
                        Subscribers
                      </li>
                      <li className="list-group-item">
                        <FontAwesomeIcon icon="eye" />{" "}
                        {separateNumber(channelIntro.statistics.viewCount)}{" "}
                        Views
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    channel: state.channel,
    error: state.error,
  };
};
export default connect(mapStateToProps, { fetchChannelIntro, clearError })(
  withRouter(ChannelIntro)
);
