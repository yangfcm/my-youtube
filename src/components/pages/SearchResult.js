import React from "react";
import { connect } from "react-redux";
import queryString from "query-string";
import SearchResultList from "../modules/SearchResultList";
import MoreButton from "../modules/MoreButton";
import Loading from "../common/Loading";
import ErrorMessage from "../common/ErrorMessage";

import { searchVideos, clearError } from "../../actions/app";

class SearchResult extends React.Component {
  state = {
    videos: null,
    error: null,
  };

  searchVideos = async (q) => {
    await this.props.searchVideos({ q });
    if (this.props.error) {
      this.setState({
        error: this.props.error,
      });
      return;
    }
    if (this.props.videos) {
      if (this.props.videos.items.length > 0) {
        this.setState({
          videos: {
            pageInfo: this.props.videos.pageInfo,
            items: this.props.videos.items,
            nextPageToken: this.props.videos.nextPageToken,
          },
        });
      } else {
        this.setState({
          error: `No video found with the key word: ${q}`,
        });
      }
    }
  };

  componentDidMount = () => {
    const { q } = queryString.parse(this.props.location.search);
    this.searchVideos(q);
  };

  componentDidUpdate = async (prevProps) => {
    const prevq = queryString.parse(prevProps.location.search).q;
    const { q } = queryString.parse(this.props.location.search);
    if (prevq !== q) {
      this.setState({
        videos: null,
        error: null,
      });
      this.searchVideos(q);
    }
  };

  componentWillUnmount = () => {
    this.props.clearError();
  };

  fetchNextPageVideos = async () => {
    const { q } = queryString.parse(this.props.location.search);
    const { nextPageToken } = this.props.videos;
    await this.props.searchVideos({ q }, nextPageToken);
    this.setState((state, props) => {
      return {
        videos: {
          pageInfo: props.videos.pageInfo,
          items: state.videos.items.concat(props.videos.items),
          nextPageToken: props.videos.nextPageToken,
        },
      };
    });
  };

  render() {
    return (
      <React.Fragment>
        {!this.state.error && !this.state.videos && <Loading />}
        {this.state.error && <ErrorMessage message={this.state.error} />}
        {!this.state.error && this.state.videos && (
          <div className="d-flex justify-content-center">
            <div className="col-lg-9">
              <h3 className="mb-3">
                Search results with keyword:{" "}
                <span className="text-danger">
                  {queryString.parse(this.props.location.search).q}
                </span>
              </h3>
              <SearchResultList searchResultList={this.state.videos.items} />

              {this.state.videos.nextPageToken && (
                <div style={{ width: "50%", margin: "0 auto" }}>
                  <MoreButton onClickMore={this.fetchNextPageVideos} />
                </div>
              )}
            </div>
          </div>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    videos: state.videos,
    error: state.error,
  };
};
export default connect(mapStateToProps, { searchVideos, clearError })(
  SearchResult
);
