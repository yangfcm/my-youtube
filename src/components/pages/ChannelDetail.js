import React from "react";
import { Switch, Route, Redirect } from "react-router-dom";
import Banner from "../layout/Banner";
import Menu from "../layout/Menu";

import ChannelIntro from "./ChannelIntro";
import ChannelVideos from "./ChannelVideos";
import ChannelPlaylist from "./ChannelPlaylist";
import ChannelBanner from "../layout/ChannelBanner";

class ChannelDetail extends React.Component {
  render() {
    const { path, url } = this.props.match;
    const channelMenuItems = [
      { link: `${url}/videos`, exact: false, name: "Videos", icon: "video" },
      { link: `${url}/playlist`, exact: false, name: "Playlist", icon: "list" },
      {
        link: `${url}/intro`,
        exact: false,
        name: "Channel Intro",
        icon: "info-circle",
      },
    ];
    return (
      <div>
        <ChannelBanner channelId={this.props.match.params.id} />
        <Menu menuItems={channelMenuItems} />
        <Switch>
          <Route path={`${path}/videos`}>
            <ChannelVideos channelId={this.props.match.params.id} />
          </Route>
          <Route path={`${path}/playlist`}>
            <ChannelPlaylist channelId={this.props.match.params.id} />
          </Route>
          <Route path={`${path}/intro`}>
            <ChannelIntro channelId={this.props.match.params.id} />
          </Route>
          <Route path={`${path}`} exact>
            <Redirect to={`${url}/videos`} />
          </Route>
          <Route path={`${path}/*`}>
            <Redirect to="/not-found" />
          </Route>
        </Switch>
      </div>
    );
  }
}

export default ChannelDetail;
