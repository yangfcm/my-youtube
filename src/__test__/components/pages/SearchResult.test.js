import React from "react";
import { shallow } from "enzyme";
import { SearchResult } from "components/pages/SearchResult";
import SearchResultList from "components/modules/SearchResultList";
import MoreButton from "components/modules/MoreButton";
import Loading from "components/common/Loading";
import ErrorMessage from "components/common/ErrorMessage";
import {
  keyword,
  searchResultEmptyResponse,
  searchResultResponse,
  searchErrorResponse,
} from "__test__/fixtures/search";

describe("Test SearchResult component", () => {
  let wrapper;
  const mockSearchVideos = jest.fn();
  const mockClearError = jest.fn();
  const location = {
    search: `?q=${keyword}`,
  };

  afterEach(() => {
    wrapper.unmount();
  });

  describe("When component is initially rendered", () => {
    beforeEach(() => {
      wrapper = shallow(
        <SearchResult
          searchVideos={mockSearchVideos}
          clearError={mockClearError}
          location={location}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render Loading component", () => {
      expect(wrapper.find(Loading).exists()).toBe(true);
    });

    it("should fetch search result", () => {
      expect(mockSearchVideos).toHaveBeenCalledWith({ q: keyword });
    });
  });

  describe("When error occurs", () => {
    beforeEach(() => {
      wrapper = shallow(
        <SearchResult
          searchVideos={mockSearchVideos}
          clearError={mockClearError}
          location={location}
          error={searchErrorResponse}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render ErrorMessage component", () => {
      expect(wrapper.find(ErrorMessage).exists()).toBe(true);
    });
  });

  describe("When search result is fetched", () => {
    beforeEach(() => {
      wrapper = shallow(
        <SearchResult
          searchVideos={mockSearchVideos}
          clearError={mockClearError}
          location={location}
          error={null}
          videos={searchResultResponse}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render SearchResultList component and pass data to it", () => {
      const searchResultList = wrapper.find(SearchResultList);
      expect(searchResultList.exists()).toBe(true);
      expect(searchResultList.prop("searchResultList")).toEqual(
        searchResultResponse.items
      );
    });

    it("should render more button and respond to click", () => {
      const moreButton = wrapper.find(MoreButton);
      expect(moreButton.exists()).toBe(true);
      moreButton.prop("onClickMore")();
      expect(mockSearchVideos).toHaveBeenCalledWith(
        { q: keyword },
        searchResultResponse.nextPageToken
      );
    });

    it("should update component when search keyword is changed", () => {
      wrapper.setProps({
        location: {
          search: "?q=new",
        },
      });
      expect(mockSearchVideos).toHaveBeenCalledWith({ q: "new" });
    });
  });

  describe("When no result is fetched", () => {
    beforeEach(() => {
      wrapper = shallow(
        <SearchResult
          searchVideos={mockSearchVideos}
          clearError={mockClearError}
          location={location}
          error={null}
          videos={searchResultEmptyResponse}
        />
      );
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render ErrorMessage component", () => {
      expect(wrapper.find(ErrorMessage).exists()).toBe(true);
      expect(wrapper.find(ErrorMessage).prop("error")).toEqual({
        displayMessage: `No video found with the key word: ${keyword}`,
      });
    });
  });
});
