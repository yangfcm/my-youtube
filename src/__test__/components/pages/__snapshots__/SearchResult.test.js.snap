// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SearchResult component When component is initially rendered should render component 1`] = `
<Fragment>
  <Loading />
</Fragment>
`;

exports[`Test SearchResult component When error occurs should render component 1`] = `
<Fragment>
  <ErrorMessage
    error={
      Object {
        "response": Object {
          "data": Object {
            "error": Object {
              "code": 400,
              "errors": Array [
                Object {
                  "domain": "youtube.part",
                  "location": "part",
                  "locationType": "parameter",
                  "message": "'snippe'",
                  "reason": "unknownPart",
                },
              ],
              "message": "'snippe'",
            },
          },
        },
      }
    }
  />
</Fragment>
`;

exports[`Test SearchResult component When no result is fetched should render component 1`] = `
<Fragment>
  <ErrorMessage
    error={
      Object {
        "displayMessage": "No video found with the key word: google dev",
      }
    }
  />
</Fragment>
`;

exports[`Test SearchResult component When search result is fetched should render component 1`] = `
<Fragment>
  <div
    className="d-flex justify-content-center"
  >
    <div
      className="col-lg-9"
    >
      <h3
        className="mb-3"
      >
        Search results with keyword:
         
        <span
          className="text-danger"
        >
          google dev
        </span>
      </h3>
      <SearchResultList
        searchResultList={
          Array [
            Object {
              "etag": "XYpvNChnzsvfjhXgkodwf1QKZsA",
              "id": Object {
                "channelId": "UC_x5XG1OV2P6uZZ5FSM9Ttw",
                "kind": "youtube#channel",
              },
              "kind": "youtube#searchResult",
              "snippet": Object {
                "channelId": "UC_x5XG1OV2P6uZZ5FSM9Ttw",
                "channelTitle": "Google Developers",
                "description": "The Google Developers channel features talks from events, educational series, best practices, tips, and the latest updates across our products and platforms.",
                "liveBroadcastContent": "none",
                "publishTime": "2007-08-23T00:34:43Z",
                "publishedAt": "2007-08-23T00:34:43Z",
                "thumbnails": Object {
                  "default": Object {
                    "url": "https://yt3.ggpht.com/-Fgp8KFpgQqE/AAAAAAAAAAI/AAAAAAAAAAA/Wyh1vV5Up0I/s88-c-k-no-mo-rj-c0xffffff/photo.jpg",
                  },
                  "high": Object {
                    "url": "https://yt3.ggpht.com/-Fgp8KFpgQqE/AAAAAAAAAAI/AAAAAAAAAAA/Wyh1vV5Up0I/s800-c-k-no-mo-rj-c0xffffff/photo.jpg",
                  },
                  "medium": Object {
                    "url": "https://yt3.ggpht.com/-Fgp8KFpgQqE/AAAAAAAAAAI/AAAAAAAAAAA/Wyh1vV5Up0I/s240-c-k-no-mo-rj-c0xffffff/photo.jpg",
                  },
                },
                "title": "Google Developers",
              },
            },
            Object {
              "etag": "LVcDd4wjf6nftSQTOeuSrsWPisQ",
              "id": Object {
                "channelId": "UCnUYZLuoy1rq1aVMwx4aTzw",
                "kind": "youtube#channel",
              },
              "kind": "youtube#searchResult",
              "snippet": Object {
                "channelId": "UCnUYZLuoy1rq1aVMwx4aTzw",
                "channelTitle": "Google Chrome Developers",
                "description": "Making the web more awesome.",
                "liveBroadcastContent": "none",
                "publishTime": "2012-04-24T00:05:52Z",
                "publishedAt": "2012-04-24T00:05:52Z",
                "thumbnails": Object {
                  "default": Object {
                    "url": "https://yt3.ggpht.com/-9X-_QJu0RvQ/AAAAAAAAAAI/AAAAAAAAAAA/X3cPcZ5icto/s88-c-k-no-mo-rj-c0xffffff/photo.jpg",
                  },
                  "high": Object {
                    "url": "https://yt3.ggpht.com/-9X-_QJu0RvQ/AAAAAAAAAAI/AAAAAAAAAAA/X3cPcZ5icto/s800-c-k-no-mo-rj-c0xffffff/photo.jpg",
                  },
                  "medium": Object {
                    "url": "https://yt3.ggpht.com/-9X-_QJu0RvQ/AAAAAAAAAAI/AAAAAAAAAAA/X3cPcZ5icto/s240-c-k-no-mo-rj-c0xffffff/photo.jpg",
                  },
                },
                "title": "Google Chrome Developers",
              },
            },
            Object {
              "etag": "gnr-mSrcXXuX3s59GvE0ArOWaUg",
              "id": Object {
                "kind": "youtube#playlist",
                "playlistId": "IDZyZMYn3H0",
              },
              "kind": "youtube#searchResult",
              "snippet": Object {
                "channelId": "UCsBjURrPoezykLs9EqgamOA",
                "channelTitle": "Fireship",
                "description": "What does it take to become a Google Developer Expert? Learn about my own experience becoming a GDE with practical advice for working towards this goal.",
                "liveBroadcastContent": "none",
                "publishTime": "2019-05-23T21:06:17Z",
                "publishedAt": "2019-05-23T21:06:17Z",
                "thumbnails": Object {
                  "default": Object {
                    "height": 90,
                    "url": "https://i.ytimg.com/vi/IDZyZMYn3H0/default.jpg",
                    "width": 120,
                  },
                  "high": Object {
                    "height": 360,
                    "url": "https://i.ytimg.com/vi/IDZyZMYn3H0/hqdefault.jpg",
                    "width": 480,
                  },
                  "medium": Object {
                    "height": 180,
                    "url": "https://i.ytimg.com/vi/IDZyZMYn3H0/mqdefault.jpg",
                    "width": 320,
                  },
                },
                "title": "How to Become a Google Developer Expert (GDE)",
              },
            },
            Object {
              "etag": "fUy83ga9B0qPSeFNWMj7RhWuLUY",
              "id": Object {
                "kind": "youtube#video",
                "videoId": "H89hKw06iWs",
              },
              "kind": "youtube#searchResult",
              "snippet": Object {
                "channelId": "UCnUYZLuoy1rq1aVMwx4aTzw",
                "channelTitle": "Google Chrome Developers",
                "description": "Join us for day 1 of web.dev LIVE as we head to the Americas timezone to provide helpful content on Performance and Discovery through Search. Enjoy the ...",
                "liveBroadcastContent": "none",
                "publishTime": "2020-06-30T16:00:10Z",
                "publishedAt": "2020-06-30T16:00:10Z",
                "thumbnails": Object {
                  "default": Object {
                    "height": 90,
                    "url": "https://i.ytimg.com/vi/H89hKw06iWs/default.jpg",
                    "width": 120,
                  },
                  "high": Object {
                    "height": 360,
                    "url": "https://i.ytimg.com/vi/H89hKw06iWs/hqdefault.jpg",
                    "width": 480,
                  },
                  "medium": Object {
                    "height": 180,
                    "url": "https://i.ytimg.com/vi/H89hKw06iWs/mqdefault.jpg",
                    "width": 320,
                  },
                },
                "title": "web.dev LIVE 2020: Day One",
              },
            },
            Object {
              "etag": "jK7Ncy4_CI-PKWWCXXrJSD-cQBs",
              "id": Object {
                "kind": "youtube#video",
                "videoId": "oHBFD8cfXho",
              },
              "kind": "youtube#searchResult",
              "snippet": Object {
                "channelId": "UCaO6VoaYJv4kS-TQO_M-N_g",
                "channelTitle": "ClÃ©ment Mihailescu",
                "description": "This is what a real day in the life of a Google Software Engineer looks like. What I worked on at Google: https://youtu.be/Az_PtvMt5_Y Song at the beginning of ...",
                "liveBroadcastContent": "none",
                "publishTime": "2019-09-24T14:00:09Z",
                "publishedAt": "2019-09-24T14:00:09Z",
                "thumbnails": Object {
                  "default": Object {
                    "height": 90,
                    "url": "https://i.ytimg.com/vi/oHBFD8cfXho/default.jpg",
                    "width": 120,
                  },
                  "high": Object {
                    "height": 360,
                    "url": "https://i.ytimg.com/vi/oHBFD8cfXho/hqdefault.jpg",
                    "width": 480,
                  },
                  "medium": Object {
                    "height": 180,
                    "url": "https://i.ytimg.com/vi/oHBFD8cfXho/mqdefault.jpg",
                    "width": 320,
                  },
                },
                "title": "A REAL Day in the Life of a Google Software Engineer",
              },
            },
          ]
        }
      />
      <div
        style={
          Object {
            "margin": "0 auto",
            "width": "50%",
          }
        }
      >
        <MoreButton
          onClickMore={[Function]}
        />
      </div>
    </div>
  </div>
</Fragment>
`;
